### PR TITLE
rose macro: fix trigger transform order-sensitivity bug

### PR DESCRIPTION
**Pull request by benfitzpatrick**
The `trigger` mechanism in transform macro mode (`rose macro`, and `rose edit` for trigger error checking) can sometimes (rarely?) be incorrect.

The code currently does:
```python
        for var_id in self.trigger_family_lookup:
            self.update(var_id, config, meta_config)
```
where `self.update` launches a cascading check down the trigger chain from `var_id`.

These cascading chains will overlap, and every setting will be checked at least once, but the unordered nature of the `for` loop means that in some particular cases the ignored status of a parent will not get properly transmitted to an ignored status for a child.

The new code loops over ids based on their maximum rank in the trigger hierarchy, from lowest to highest. This means that any parent should propagate their status properly (reliably!) to their children.

I've added a test that fails on current master and succeeds on this branch. The test is very sensitive to the exact contents of the metadata configuration, which is why it has settings that look like unnecessary junk - once removed or renamed, the test will no longer fail!

@matthewrmshin, @arjclark please review.

I've checked the change is null with respect to the latest UM trunk, and will check vs our suites to get a handle on the effects.

### CODE REVIEW

File: lib/python/rose/macros/trigger.py [hunk 1]
```diff
@@ -262,6 +267,24 @@ def update(self, var_id, config_data, meta_config):
             id_stack.pop(0)
         return update_id_list
 
+    def _get_ranked_trigger_ids(self):
+        """Return trigger ids with their maximum trigger chain depth."""
```
Comment (matthewrmshin) [inline lib/python/rose/macros/trigger.py]: Probably worth moving your comment in line 94 down here?

File: lib/python/rose/macros/trigger.py [hunk 2]
```diff
@@ -262,6 +267,24 @@ def update(self, var_id, config_data, meta_config):
...
+        stack = [(id_, 0) for id_ in self.trigger_family_lookup]
```
Comment (matthewrmshin) [inline lib/python/rose/macros/trigger.py]: May be worth adding a comment on what's in each element of `stack`?

**Comment by matthewrmshin** [comment]
Logic looks OK 1. Test OK. Some comments on more comments.

**Comment by benfitzpatrick** [comment]
Comment to comment on comment changes addressing your comments on more comments

**Comment by matthewrmshin** [comment]
@arjclark please review 2.

**Comment by arjclark** [comment]
Looks OK to me. No problems from the test-battery in my environment.

**Pull request merged**